pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -178,8 +178,30 @@ func (p *DNSProxy) checkRestored(endpointID uint64, destPort uint16, destIP stri
 		return false
 	}
 
+	dest, err := restore.ParseRuleIPOrCIDR(destIP)
+	if err != nil || !dest.IsAddr() {
+		return false
+	}
```

Comment:
It's an over-cautious sanity check. `destIP` being a packet's destination it is always an IP address, never a CIDR.

```
destAddr, err := netip.ParseAddr(destIP)
if err != nil {
    return false
}
dest := restore.RuleIPOrCIDR(netip.PrefixFrom(destAddr, 0xff)) // should introduce a function for this
```

should express the intent better. As an aside, with the new immutable types in `net/netip` the use of strings could be phased out all over the place, but it's a huge amount of work.

Reply:
The lookup is done with both `dest` and `dest.ToSingleCIDR()` because the set of rule IPs can contain both bare addresses (such as from resolution of `toEndpoint` rules) and CIDRs. I do these two checks as a fast lookup that covers most use cases before looping all over the set and (usually) rejecting the request.